File: 100-The 3n + 1 Problem/band.cpp
#include <iostream>
#include <algorithm>
using namespace std;


int algo(int a);
int band(int, int);


int main()
{
    int i = 0, j = 0;
    while(cin>>i>>j)
    {
        cout<<i<<" "<<j<<" "<<band(i, j)<<endl;
    }

    return 0;
}


int algo(int a)
{
    int cycle = 1;

    while (a!=1)
    {
        a = (a % 2 != 0) ? (a * 3 + 1) : (a = a / 2);

        cycle++;
    }

    return cycle;
}


int band(int val_one, int val_two)
{
    int minn = (val_one < val_two) ? val_one : val_two;   int maxx = (val_one > val_two) ? val_one : val_two;   int maxCycle = -1, cycle = 0;

    for(int a = minn; a <= maxx; a++)
    {
        cycle = algo(a);    maxCycle = (cycle < maxCycle) ? maxCycle : cycle;
    }

    return maxCycle;
}

File: 100-The 3n + 1 Problem/band.py
import sys
import functools


@functools.lru_cache(None)
def band(val_one, val_two):
    val_one, val_two = min(val_one, val_two), max(val_one, val_two)
    
    return max(algo(a) for a in range(val_one, val_two+1))


@functools.lru_cache(None)
def algo(a):
    cycle = 1

    while 1 != a:
        if a % 2 != 0:
            a = 3*a + 1
        else:
            a = a // 2
            
        cycle += 1
    
    return cycle


if __name__ == '__main__':    
    for a in sys.stdin:
        val_one, val_two = map(int, a.split())
        print(val_one, val_two, band(val_one, val_two))
        
    exit(0)